ethers-providers/src/rpc/transports/middleware.rs

```diff
@@ -0,0 +1,176 @@
+use bytes::Bytes;
+use std::collections::HashMap;
+
+use crate::{
+    rpc::transports::http::{ClientError, Provider},
+    HttpClientError::{MiddlewareError, ReqwestError},
+};
+use anyhow::anyhow;
+use http::response::Builder;
+use reqwest::{Body, Response, StatusCode};
+use reqwest_chain::Chainer;
+use reqwest_middleware::Error;
+use url::Url;
+
+/// Middleware for switching between providers on failures
+pub struct SwitchProviderMiddleware {
+    /// Providers for the url
+    pub providers: Vec<Url>,
+}
+
+#[derive(Default, Debug)]
+pub struct LocalState {
+    pub active_provider_index: usize,
+    pub prev_stat: HashMap<usize, Option<ClientError>>,
+}
+
+impl SwitchProviderMiddleware {
+    pub fn _new(providers: Vec<Url>) -> Self {
+        Self { providers }
+    }
+}
+
+#[async_trait::async_trait]
+impl Chainer for SwitchProviderMiddleware {
+    type State = LocalState;
+
+    async fn chain(
+        &self,
+        result: Result<reqwest::Response, Error>,
+        _state: &mut Self::State,
+        request: &mut reqwest::Request,
+    ) -> Result<Option<reqwest::Response>, Error> {
+        let mut next_state = |client_error: Option<ClientError>| {
+            let active_index = _state.active_provider_index;
+            _state.prev_stat.insert(active_index, client_error);
+            let next_index = _state.active_provider_index + 1;
+            if next_index >= self.providers.len() {
+                return Err(anyhow!("Exiting Middleware"))?;
```

Comment:
Let's return a client error here

Reply:
You can add a trace log that lets the consumer know the providers have been exhausted.